File: implementations/java-springboot/examples/src/main/java/com/example/accountservice/card/infrastructure/persistence/CardMapper.java
package com.example.accountservice.card.infrastructure.persistence;

import com.example.accountservice.card.domain.Card;
import java.time.YearMonth;

/**
 * A class dedicated to converting between Card (pure domain) and CardJpaEntity (JPA mapping). Used
 * only internally within CardRepositoryImpl — the Domain/Application layers have no knowledge of
 * this class.
 */
final class CardMapper {

    private CardMapper() {}

    static Card toDomain(CardJpaEntity entity) {
        return Card.reconstitute(
                entity.getCardId(),
                entity.getAccountId(),
                entity.getOwnerId(),
                entity.getBrand(),
                entity.getStatus(),
                entity.getCreatedAt(),
                toYearMonth(entity.getLastStatementSentMonth()));
    }

    /** Creates a new entity (no PK, to be inserted) for a newly issued Card. */
    static CardJpaEntity toNewEntity(Card card) {
        return new CardJpaEntity(
                null,
                card.getCardId(),
                card.getAccountId(),
                card.getOwnerId(),
                card.getBrand(),
                card.getStatus(),
                card.getCreatedAt(),
                toColumn(card.getLastStatementSentMonth()));
    }

    /**
     * Applies the domain Card's latest state onto an existing entity (preserving its PK) — to be
     * updated.
     */
    static CardJpaEntity updateEntity(CardJpaEntity entity, Card card) {
        entity.applyMutableState(card.getStatus(), toColumn(card.getLastStatementSentMonth()));
        return entity;
    }

    private static YearMonth toYearMonth(String column) {
        return column == null ? null : YearMonth.parse(column);
    }

    private static String toColumn(YearMonth yearMonth) {
        return yearMonth == null ? null : yearMonth.toString();
    }
}

File: implementations/kotlin-springboot/examples/src/main/kotlin/com/example/accountservice/card/infrastructure/persistence/CardMapper.kt
package com.example.accountservice.card.infrastructure.persistence

import com.example.accountservice.card.domain.Card

/**
 * Object dedicated to converting between Card (pure domain) and CardJpaEntity (JPA mapping).
 * Used only inside CardRepositoryImpl — the Domain/Application layers know nothing of this object.
 */
internal object CardMapper {
    fun toDomain(entity: CardJpaEntity): Card =
        Card.reconstitute(
            cardId = entity.cardId,
            accountId = entity.accountId,
            ownerId = entity.ownerId,
            brand = entity.brand,
            status = entity.status,
            createdAt = entity.createdAt,
            lastStatementSentMonth = entity.lastStatementSentMonth,
        )

    /** Creates a new entity for a new Card (no PK, subject to insert). */
    fun toNewEntity(card: Card): CardJpaEntity =
        CardJpaEntity(
            id = null,
            cardId = card.cardId,
            accountId = card.accountId,
            ownerId = card.ownerId,
            brand = card.brand,
            status = card.status,
            createdAt = card.createdAt,
            lastStatementSentMonth = card.lastStatementSentMonth,
        )

    /** Applies the domain Card's latest state (status/lastStatementSentMonth) onto an existing entity (PK preserved) — subject to update. */
    fun updateEntity(
        entity: CardJpaEntity,
        card: Card,
    ): CardJpaEntity {
        entity.status = card.status
        entity.lastStatementSentMonth = card.lastStatementSentMonth
        return entity
    }
}
